### PR TITLE
Fixes #11898 - Information about releases available on 'About' page

### PR DESCRIPTION
**Pull request by xprazak2**
Needs [changes](https://github.com/theforeman/theforeman.org/pull/550) on our Wiki to work. The main idea is that we load json data provided by wiki, sift through it and display what we want. I added latest released versions of core and latest version of plugins if available.

### CODE REVIEW

**Comment by timogoebel** [comment]
I like the feature, however we should be very careful with calling resources on "the web" from within the application as this may be blocked by a firewall (read: probably will be blocked on most corporate environments). I think, there should be a global setting do disable any "phone home" features.  And the feature should have a very short timeout as some firewalls may just drop the packets and not reject them.

**Comment by xprazak2** [comment]
The setting to turn off the feature is probably a good idea, I will make changes.

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -431,11 +431,19 @@ def edit_inline(object, property, options = {})
     end
   end
 
-  def documentation_url(section = "")
-    if section.empty?
-      "http://www.theforeman.org/documentation.html##{SETTINGS[:version].short}"
+  def documentation_url(section = "", version = "")
+    if version.empty?
```
Comment (adamruzicka) [inline app/helpers/application_helper.rb]: How about something like

``` ruby
version = SETTINGS[:version].short if version.empty?
if section.empty?
  "#{Wiki::URL}/documentation.html##{version}"
else
  "#{Wiki::URL}/manuals/#{version}/index.html#" + section
end
```

so you don't have the urls on two places

**Comment by unorthodoxgeek** [comment]
will this work for Satellite as well?

**Comment by domcleal** [comment]
I'm going to set this as WoC until the corresponding changes to the website are done.

**Comment by lzap** [comment]
I would like to reuse this core code in discovery to determine latest Discovery Image version. This will be used in Show discovered host page to notify user it has old image running. Can you please make sure the next code rebase:
- works with arbitrary names (I will use `discovery_image` in the JSON perhaps)
- JSON retrieval is cached in Rails cache with at least one hour (don't want to slower our UI)
- the API can be easily used from Plugins

**Comment by lzap** [comment]
Re Satellite - I think this patch should also allow possibility to easily "override" the JSON with a local file, so for downstream installations, installer could just drop the JSON somewhere and this would prevent Foreman from accessing our web page.

**Comment by lzap** [comment]
And paranoic users could also schedule a wget/curl command to fetch this JSON manually...

File: app/helpers/about_helper.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+module AboutHelper
+  def latest_version(releases, plugin_name)
+    versions = releases["releases"]["plugins"][plugin_name] rescue nil
```
Comment (ohadlevy) [inline app/helpers/about_helper.rb]: rescue [] otherwise next line fails?

Comment (ohadlevy) [inline app/helpers/about_helper.rb]: ah, didnt see the if, either way is fine.

File: app/helpers/about_helper.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    latest = versions.map { |v| Gem::Version.new(v) if Gem::Version.correct?(v) }.compact.max if versions
+    latest || _("N/A")
+  end
+
+  def latest_foreman_versions(releases)
+    versions = releases["releases"]["core"]["latest"] rescue nil
```
Comment (ohadlevy) [inline app/helpers/about_helper.rb]: extract to a method?

File: app/models/setting/general.rb [hunk 4]
```diff
@@ -22,7 +22,8 @@ def self.load_defaults
         self.set('use_gravatar', N_("Foreman will use gravatar to display user icons"), false, N_('Use Gravatar')),
         self.set('db_pending_migration', N_("Should the `foreman-rake db:migrate` be executed on the next run of the installer modules?"), true, N_('DB pending migration')),
         self.set('db_pending_seed', N_("Should the `foreman-rake db:seed` be executed on the next run of the installer modules?"), true, N_('DB pending seed')),
-        self.set('proxy_request_timeout', N_("Max timeout for REST client requests to smart-proxy"), 60, N_('Proxy request timeout'))
+        self.set('proxy_request_timeout', N_("Max timeout for REST client requests to smart-proxy"), 60, N_('Proxy request timeout')),
+        self.set('released_versions', N_("Foreman tries to find out about released versions"), false, N_('Find Foreman releases'))
```
Comment (ohadlevy) [inline app/models/setting/general.rb]: description:
Connect to Foreman website for latest version information?

File: app/services/wiki.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  URL = "http://projects.theforeman.org"
+
+  def versions
+    uri = URI.parse("#{Wiki::URL}/versions.json")
+    http = Net::HTTP.new(uri.host, uri.port)
+    request = Net::HTTP::Get.new(uri.request_uri)
+    request['Content-Type'] = 'application/json'
+    begin
+      result = Net::HTTP.start(uri.host, uri.port, :read_timeout => 1000) { |http| http.request(request) }
+      result.value
```
Comment (ohadlevy) [inline app/services/wiki.rb]: cache the response?

File: app/controllers/about_controller.rb [hunk 6]
```diff
@@ -5,5 +5,6 @@ def index
     @smart_proxies = SmartProxy.authorized(:view_smart_proxies).includes(:features)
     @compute_resources = ComputeResource.authorized(:view_compute_resources)
     @plugins = Foreman::Plugin.all
+    @releases = Wiki.new.versions if Setting[:released_versions]
```
Comment (ohadlevy) [inline app/controllers/about_controller.rb]: IMHO this should be a seperate AJAX request, so it does not block foreman.
further, I beleive that sometimes foreman itself wont allow connection to the internet, but maybe the user browser can, which in this case becomes a client side only implementation. I'm fine with this as an initial approach.

File: app/models/setting/general.rb (hunk 4 above)
Comment (ohadlevy) [inline app/models/setting/general.rb]: also, I would consider changing this to a more generic statement, it is true that currently it only compare version information, but I could see how other plugins and additional information could be used, and would prefer not to keep creating additional settings for each call home feature.

**Comment by xprazak2** [comment]
This is now aligned with PR for the wiki.
- info about updates is loaded asynchronously
- plugins can register their own source of updates in `Engine` by calling `register_update_source(MyUpdateClass)`, where `MyUpdateClass < Updates::Base`
- info about plugin updates show the latest plugin version for the current Foreman version. That means: if someone currently has Foreman 1.9 with a plugin that `requires_foreman >= 1.9` and new plugin version gets released with `requires_foreman >= 1.11`, it will not be displayed until user updates to 1.11
- update info can be retrieved from file. If file cannot be read for whatever reason, the Foreman tries to get it from web (if allowed). 
- info about updates can be disabled entirely
- I think this should work for Satellite without problems

Concerns: 
- Maybe split the file in two, one for core versions, one for plugins and for each register a separate update source?
- there are `url` and `file_name` methods that determine where to look for updates. These can be overridden by Satellite with `alias_method_chain` but maybe there is a better way?

TODO:
- [x] tests

**Comment by xprazak2** [comment]
[test]

**Comment by xprazak2** [comment]
[test]

**Comment by dLobatog** [comment]
@xprazak2 I'd favor more https://github.com/theforeman/foreman/pull/3400 's approach than this, have you taken a look at that one?

**Comment by xprazak2** [comment]
@dLobatog I missed it, but it looks really great. Closing.

**Pull request closed without merging**